jerry-core/ecma/base/ecma-module.h

```diff
@@ -16,12 +16,125 @@
 #ifndef ECMA_MODULE_H
 #define ECMA_MODULE_H
 
-#include "js-parser-internal.h"
-
 #if ENABLED (JERRY_ES2015_MODULE_SYSTEM)
 
-void ecma_module_load_modules (parser_context_t *context_p);
+#include "common.h"
+#include "ecma-globals.h"
+
+#define ECMA_MODULE_MAX_PATH 255u
+
+/**
+ * Imported or exported names, such as "a as b"
+ * Note: See https://www.ecma-international.org/ecma-262/6.0/#table-39
+ *       and https://www.ecma-international.org/ecma-262/6.0/#table-41
+ */
+typedef struct ecma_module_names
+{
+  struct ecma_module_names *next_p; /**< next linked list node */
+  ecma_string_t *imex_name_p;       /**< Import/export name of the item */
+  ecma_string_t *local_name_p;      /**< Local name of the item */
+} ecma_module_names_t;
+
+typedef struct ecma_module ecma_module_t;
+
+/**
+ * Module node to store imports / exports.
+ */
+typedef struct ecma_module_node
+{
+  struct ecma_module_node *next_p;     /**< next linked list node */
+  ecma_module_names_t *module_names_p; /**< names of the requested import/export node */
+  ecma_module_t *module_request_p;     /**< module structure of the requested module */
+} ecma_module_node_t;
+
+/**
+ * Module context containing all import and export nodes.
+ */
+typedef struct ecma_module_context
+{
+  struct ecma_module_context *parent_p;   /**< parent context */
+  ecma_module_node_t *imports_p;          /**< import item of the current context */
+  ecma_module_node_t *local_exports_p;    /**< export item of the current context */
+  ecma_module_node_t *indirect_exports_p; /**< export item of the current context */
+  ecma_module_node_t *star_exports_p;     /**< export item of the current context */
+  ecma_module_t *module_p;                /**< module request */
+} ecma_module_context_t;
+
+/**
+ * An enum identifing the current state of the module
+ */
+typedef enum
+{
+  ECMA_MODULE_STATE_INIT = 0,       /**< module is initialized */
+  ECMA_MODULE_STATE_PARSING = 1,    /**< module is currently being parsed */
+  ECMA_MODULE_STATE_PARSED = 2,     /**< module has been parsed */
+  ECMA_MODULE_STATE_EVALUATING = 3, /**< module is currently being evaluated */
+  ECMA_MODULE_STATE_EVALUATED = 4,  /**< module has been evaluated */
+} ecma_module_state_t;
+
+/**
+ * Module structure storing an instance of a module
+ */
+struct ecma_module
+{
+  struct ecma_module *next_p;         /**< next linked list node */
+  ecma_module_state_t state;          /**< state of the mode */
+  ecma_string_t *path_p;              /**< path of the module */
+  ecma_module_context_t *context_p;   /**< module context of the module */
```

Comment:
I do not like this kind of cycling reference. Why is this necessary?

Reply:
It's not necessary in the strict sense of the word, but it's very useful for not having to look up which module the current context belongs to when connecting properties.